### ISSUE TITLE
Class properties should/could be inferred from constructor parameters

### ISSUE THREAD

**Issue by Kocal** [vimeo/psalm#1703]
Hi, this issue is a _following_ of https://twitter.com/psalmphp/status/1133700224168452096.

tl;dr: I was complaining about the tons of `MissingPropertyType` issues I had in classes like this:
![image](https://user-images.githubusercontent.com/2103975/58555362-8e9da480-8219-11e9-99be-6381890f24f1.png)

I may be wrong, but I think it's common to do this. I mean, not declaring a property type with `@var ...` when a property is directly assigned in the constructor from a PHP-typed parameter.

For example, Symfony [do the same](https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/Mime/Message.php#L24-L33) since some time with _recent_ code. Also PhpStorm is dealing great with this:
![Sélection_772](https://user-images.githubusercontent.com/2103975/58556231-c7d71400-821b-11e9-9498-d9ae3ce44a40.png)

Considering [this example](https://psalm.dev/r/dabf4c2e64), when we have `Property Foo::$a does not have a declared type - consider XXX`, it would be nice for Psalm to infer `$a` type with `XXX` instead of giving a warning, no?
That would also prevent the error `Cannot determine the type of the object on the left hand side of this expression`, because for Psalm `$a` has no type.

Here is [the example with `@var`](https://psalm.dev/r/17fc6ab988), this is the behavior I expected before runing Psalm (but without the explicit `@var`).

What do you think about it?
Thanks! :)

**Comment by muglug** [comment]
I think this could be workable in the case where the constructor _only_ sets properties, and those properties have types, but I'd like to hear others' opinions before I go about implementing it.

**Comment by joshdifabio** [comment]
I personally omit type docs for props set in the constructor and have to ignore all of the resultant warnings Psalm generates.

My personal preference would be: If a prop has no type declared and that prop is assigned in the constructor, infer the type based on the constructor assignment. Any other assignments should be allowed but should be validated against that inferred type.

**Comment by SignpostMarv** [comment]
@muglug will it behave if the constructor is in a trait ?

**Comment by muglug** [comment]
I just wouldn't fill in trait property types

**Comment by SignpostMarv** [comment]
@muglug: i.e.

```php
abstract class Foo {
protected $thing;
}

trait Bar {
public function __construct(string $a);
}
trait Baz {
public function __construct(int $a);
}

class FooBar extends Foo {
use Bar;
}
class FooBaz extends Foo {
use Baz;
}
```

**Comment by muglug** [comment]
Yeah this would only kick in for

```php
class A {
  public $a;
  protected $b;
  private $c;

  public function __construct(int $a, string $b, bool $c) {
    $this->a = $a;
    $this->b = $b;
    $this->c = $c;
  }
}
```

i.e. non-abstract classes where the constructor just sets properties.

Any other behaviour in the constructor e.g.

```php
  public function __construct(int $a, string $b, bool $c) {
    $this->someMethodHere();
    $this->a = $a;
    $this->b = $b;
    $this->c = $c;
    $this->orSomeMethodHere();
  }
```

would stop Psalm from being able to safely infer a type
